src/me.rs

```diff
@@ -165,395 +172,361 @@ pub fn get_subset_predictors<T: Pixel>(
-    };
+fn full_pixel_me<T: Pixel>(
+  fi: &FrameInvariants<T>, ts: &TileStateMut<'_, T>,
+  org_region: &PlaneRegion<T>, p_ref: &Plane<T>, tile_bo: TileBlockOffset,
+  lambda: u32, cmvs: ArrayVec<[MotionVector; 7]>, pmv: [MotionVector; 2],
+  mvx_min: isize, mvx_max: isize, mvy_min: isize, mvy_max: isize,
+  bsize: BlockSize, ref_frame: RefType,
+) -> MVSearchResult {
+  let tile_mvs = &ts.mvs[ref_frame.to_index()].as_const();
+  let frame_ref =
+    fi.rec_buffer.frames[fi.ref_frames[0] as usize].as_ref().map(Arc::as_ref);
+  let predictors = get_subset_predictors(
+    tile_bo,
+    cmvs,
+    tile_mvs,
+    frame_ref,
+    ref_frame.to_index(),
+  );
 
-    let tile_mvs = &ts.mvs[ref_frame.to_index()].as_const();
-    let frame_ref = fi.rec_buffer.frames[fi.ref_frames[0] as usize]
-      .as_ref()
-      .map(Arc::as_ref);
+  let frame_bo = ts.to_frame_block_offset(tile_bo);
+  let po = frame_bo.to_luma_plane_offset();
+  fullpel_diamond_me_search(
+    fi,
+    po,
+    org_region,
+    p_ref,
+    &predictors,
+    fi.sequence.bit_depth,
+    pmv,
+    lambda,
+    mvx_min,
+    mvx_max,
+    mvy_min,
+    mvy_max,
+    bsize,
+  )
+}
 
-    let mut predictors = get_subset_predictors::<T>(
-      tile_bo_adj,
-      pmvs.iter().cloned().filter_map(identity).collect(),
-      tile_mvs,
-      frame_ref,
-      ref_frame.to_index(),
-    );
+fn sub_pixel_me<T: Pixel>(
+  fi: &FrameInvariants<T>, po: PlaneOffset, org_region: &PlaneRegion<T>,
+  p_ref: &Plane<T>, lambda: u32, pmv: [MotionVector; 2], mvx_min: isize,
+  mvx_max: isize, mvy_min: isize, mvy_max: isize, bsize: BlockSize,
+  use_satd: bool, best: &mut MVSearchResult, ref_frame: RefType,
+) {
+  subpel_diamond_me_search(
+    fi,
+    po,
+    org_region,
+    p_ref,
+    fi.sequence.bit_depth,
+    pmv,
+    lambda,
+    mvx_min,
+    mvx_max,
+    mvy_min,
+    mvy_max,
+    bsize,
+    use_satd,
+    best,
+    ref_frame,
+  );
+}
 
-    for predictor in &mut predictors {
-      predictor.row >>= 1;
-      predictor.col >>= 1;
-    }
+fn me_ss2<T: Pixel>(
+  fi: &FrameInvariants<T>, ts: &TileStateMut<'_, T>,
+  pmvs: &[Option<MotionVector>; 3], tile_bo_adj: TileBlockOffset,
+  rec: &ReferenceFrame<T>, global_mv: [MotionVector; 2], lambda: u32,
+  mvx_min: isize, mvx_max: isize, mvy_min: isize, mvy_max: isize,
+  bsize: BlockSize, ref_frame: RefType,
+) -> MVSearchResult {
+  let frame_bo_adj = ts.to_frame_block_offset(tile_bo_adj);
+  let po = PlaneOffset {
+    x: (frame_bo_adj.0.x as isize) << BLOCK_TO_PLANE_SHIFT >> 1,
+    y: (frame_bo_adj.0.y as isize) << BLOCK_TO_PLANE_SHIFT >> 1,
+  };
+  let org_region =
+    &ts.input_hres.region(Area::StartingAt { x: po.x, y: po.y });
+
+  let tile_mvs = &ts.mvs[ref_frame.to_index()].as_const();
+  let frame_ref =
+    fi.rec_buffer.frames[fi.ref_frames[0] as usize].as_ref().map(Arc::as_ref);
+
+  let mut predictors = get_subset_predictors::<T>(
+    tile_bo_adj,
+    pmvs.iter().cloned().filter_map(identity).collect(),
```

Comment:
Also here (filter_map)

Reply:
Collecting the non None values of an array into an ArrayVec<[MotionVector, ?]>. This code is only marked new because I removed a trait thus decreasing indentation. That has been intentionally put in a separate commit to make it easier to review.

Reply:
Thanks, for some reason it looked new when I was looking at the diff.